src/mapreduce.jl

```diff
@@ -159,27 +159,46 @@ end
 (-)(A::DArray, x::Number) = A .- x
 (-)(x::Number, A::DArray) = x .- A
 
-map_localparts(f::Callable, d::DArray) = DArray(i->f(localpart(d)), d)
-map_localparts(f::Callable, d1::DArray, d2::DArray) = DArray(d1) do I
-    f(localpart(d1), localpart(d2))
+function map_localparts(f::Callable, d::DArray)
+    r = Base.asyncmap!(similar(procs(d), Future), procs(d)) do p
+        remotecall_wait(p) do
+            f(localpart(d))
+        end
+    end
+    return DArray(r)
```

Comment:
Thanks for the comments. I think you are right that this is not the right approach. It is pretty inefficient to have a `DArray` with only a single element on each worker so I don't think it's work it to add complexity to the existing constructors to support this. I can see some other solutions. Basically, `map_localparts` would continue to assume that the shape was unchanged and then have a similar `reduce_localparts` and `mapreduce_localparts` which returned an `Array` instead of a `DArray`. I'm wondering if it is worth overloading `localpart` such that we could have `localpart(mapreduce, abs2, +, A)`.